VHAR-5571 Materialien suunnittelun tallennuksesta tulee tietokantapoikkeuksia jos valittuna kopiointi tuleville hoitokausille

jarnovayrynen opened the pull request.

jarnovayrynen committed:
VHAR-5571 Materialien suunnittelun tallennuksesta tulee tietokantapoikkeuksia

Jos valittuna kopiointi tuleville hoitokausille, oli rivien pvm eri muodossa kuin kannasta nostaessa. Tämä aiheutti sen ettei tunnistettu kannassa olevia rivejä päivitykseksi, vaan niistä tuli inserttejä ja UNIQUE–rajoite paukkui.

jarnovayrynen commented on src/cljs/harja/views/hallinta/integraatiotilanne.cljs:
```diff
@@ -20,8 +20,8 @@
           [integraatioloki/integraatioloki])
 
-        "Sonjajonot"
+        "JMS-jonot"
         :sonjajonot
         (when (oikeudet/hallinta-integraatiotilanne-sonjajonot)
```
Väärin kutsua ITFM ja Sonjajonoja "Sonjajonoiksi", koska vain toinen on Sonjajono. Siksi JMS-jonot

jarnovayrynen commented on tietokanta/testidata/hoitoluokittaiset_materiaalit.sql:
```diff
@@ -8,7 +8,7 @@ VALUES
    (SELECT id FROM sopimus WHERE urakka = (SELECT id FROM urakka WHERE nimi='Oulun alueurakka 2014-2019') AND paasopimus IS null), NOW(),
    (SELECT id FROM kayttaja WHERE kayttajanimi='jvh')),
   ('20171001', '20180930', 1000, (SELECT id FROM materiaalikoodi WHERE nimi = 'Hiekoitushiekka'),
-   (SELECT id FROM urakka WHERE nimi='Oulun alueurakka 2014-2019'),
```
Tässä oli vain testidatan tekovirhe. Kajaanin urakkaan piti tehdä sama kuin Ouluun, mutta urakka id tuli Oulusta. Soppari Kajaanista.

jarnovayrynen commented on src/clj/harja/palvelin/palvelut/materiaalit.clj:
```diff
@@ -111,10 +109,8 @@
-        (log/debug "Päivitetäänkö myös tulevilta? " tulevat-hoitokaudet-mukana?)
-
-        (let [vanhat-materiaalit (get vanhat-materiaalit hoitokausi)
+        (let [hoitokausi (mapv pvm/paivan-alussa hoitokausi) ;; muuten tulee mismatch kellonajan ja kannan kanssa, VHAR-5571
```
Tämä on se rivi mikä korjaa ongelman. Frontilta, kun käytettiin tuleville-hoitokausille? flagiä, tuli hoitokauden loppupvm muodossa 30.9.2021 klo 23:59:59, kun kannasta ne nousee datena 30.9.2021 klo 00:00. Tämä ei heijastele minnekään muualle eli pitäisi olla turvallinen korjaus.

K1ll3rF0x replied:
Tämä tuntuu minusta vähän laastarilta korjaukseksi. Mikä on syy, että tulee tuo hoitokauden loppupvm väärässä muodossa, kun Harjan omat apufunktiotkin palauttaa hoitokauden alkupvm ja loppupvm oikeassa muodossa. Onkohan ongelman alkulähteellä käytetty jotain omaa apufunktiota, joka tekee asiat eri tavalla?

(harja.pvm/paivan-alussa (java.util.Date. 1432432432325))
=> #inst"2015-05-23T21:00:00.000-00:00"
(harja.pvm/hoitokauden-loppupvm 2021)
=> #inst"2021-09-29T21:00:00.000-00:00"
(harja.pvm/paivamaaran-hoitokausi (java.util.Date. 1432432432325))
=> [#inst"2014-09-30T21:00:00.000-00:00" #inst"2015-09-29T21:00:00.000-00:00"]

jarnovayrynen committed:
Käsittele mat. suunn. rivejä pvm:nä myös frontissa

Lisää paremmat kommentit koodiin mistä on kyse

jarnovayrynen committed:
Merge branch 'develop' of github.com:finnishtransportagency/harja into VHAR-5571

jarnovayrynen committed:
Korjaa materiaalin käytön uniikkiusvaatimusta

Oltava uniikki jos ei poistettu

jarnovayrynen committed:
Tee hoitokausittain ryhmittelyn korjaus backendissä

Testailujen jälkeen kävi ilmi että paras paikka korjata suunniteltujen materiiaalien pvm:n mukaan ryhmittely on tallennuksen alkuvaiheessa backendissä. Frontissa ryhmittelyt jne. toimivat oikein frontin käyttämien hoitoaikojen mukaan, jotka sisältävät loppuppvm aikaleiman 23:59:59. Tallennusvaiheessa kuitenkin normalisoidaan kaikki vain pvm:n mukaan niin ei tule virheitä updateissa tai deleteissä.
